src/Waives.Http/ExceptionHandlingRequestSender.cs

```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Waives.Http
+{
+    internal class ExceptionHandlingRequestSender : IHttpRequestSender
+    {
+        private readonly IHttpRequestSender _wrappedRequestSender;
+
+        public ExceptionHandlingRequestSender(IHttpRequestSender wrappedRequestSender)
+        {
+            _wrappedRequestSender = wrappedRequestSender ?? throw new ArgumentNullException(nameof(wrappedRequestSender));
+        }
+
+        /// <summary>
+        /// Send a request to the wrapped IHttpRequestSender and make sure that any exception is transformed
+        /// into a WaivesApiException.
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        public async Task<HttpResponseMessage> Send(HttpRequestMessage request)
+        {
+            try
+            {
+                return await _wrappedRequestSender.Send(request).ConfigureAwait(false);
+            }
+            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
+            {
+                // Either TaskCanceledException or OperationCanceledException may be thrown by HttpClient if a response
+                // is not received before the HttpClient's TimeOut expires
+                throw new WaivesApiException($"{request.Method} request to {request.RequestUri} timed-out (client-side)");
+            }
+            catch (Exception e)
```

Comment:
💪 Can this be a more specific exception, like an `HttpRequestException`? [According to the docs](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.sendasync?view=netstandard-2.0), this will also catch _at least_ an `ArgumentNullException` (if the request message is null) and an `InvalidOperationException` (if the request message has already been sent).

Reply:
I'm not sure what we'd do differently in these cases though? The intent is to only throw `WaivesApiException`s from here, so the retry policy doesn't need to know about different exception types.